### PR TITLE
Initial/experimental support for @ping in chat

### PR DESCRIPTION
**Pull request by clifffrey**
Right now, if any chat message you can see contains "@username" or "@ username", or if someone whispers at you, then the following effects happen:
1) if you are at 100% health, you take 1 damage
2)A popup appears near the top-center of the screen with the message contents, which disappears after 10 seconds (the one popup will contain multiple messages in it if you get many pings at a time)
3) A 'new_objective' sound plays at 0.6 volume

I considered also drawing an explosion at the location of the player or something, but it seems unnecessary to me compared to the popup.

Right now, this needs to be manually enabled under Fish -> Config for each user that wants to see the ping effects.

I'm hoping to eventually enable something like this by default (although people could still disable it), but I need feedback about what the right effect is first (and if we need rate limiting or something first).

### Tested Changes:
- [x] I've tested the changes locally or with people.
- [ ] I've not tested the changes.

### CODE REVIEW

**Review by sciqu — APPROVED** [review]
If a player enables then disables the slider, they still get pings. Suspect it's because they remain in "want_pings" table (as false) and it just checks if a player is in that table at all?

big-explosion is too big imo. Tried uranium-cannon-explosion (too inconsistent) and orange-circle-with-arrow (a permanent entity unlike explosions). land-mine-explosion seems good. The GUI popup is annoying because it shifts every other GUI including the massive GUI in top left, for a second.

**Comment by sciqu** [comment]
A sound effect will be cool

**Comment by clifffrey** [comment]
The problem with not doing the GUI change is that if a player is in map mode (like most spectators are likely to be, for instance), they will not see anything.  Thus I think that some form of popup is a good idea.  We could use player.gui.screen instead to just put it at an absolute position.... I am very much open to suggestions.

I essentially always play on mute, so I don't end up thinking about sound effects....

**Comment by penguinencounter** [comment]
Definitely go with `player.gui.screen`. Easiest positioning option is centered (`auto_center` etc) but absolute positioning based on resolution and scale is doable.

**Comment by clifffrey** [comment]
I'd love if someone would propose what the actual Frame should look like... It feels strange to me just having a label saying PING... but I guess this is just the first version...

**Comment by ness056** [comment]
Maybe instead of a visual thing it would be better to play a sound, something like the discord ping sound.

**Comment by penguinencounter** [comment]
@clifffrey who sent the message, contents of the message

mockup:
cliff_build says @penguinencounter please stop dropping deconstruction planners on the iron belt; 30 seconds ago - click to dismiss

**Comment by clifffrey** [comment]
> If a player enables then disables the slider, they still get pings. Suspect it's because they remain in "want_pings" table (as false) and it just checks if a player is in that table at all?

Indeed, thank you. Fixed.

**Comment by clifffrey** [comment]
Interesting... When looking for code that could play a sound, I found `utils.alert`, which does a very similar thing, including playing a sound, and has a little progress bar for when it auto-disappears.
image

I appreciate the mock-up of what the popup should possibly look like.  It is actually hard for me to implement that, because right now, I just get the whole message, i.e. "(north) cliff_build [to south]: my message" and I would need to either search for ":" and remove whatever is before that, or do more code refactoring, and I decided to just leave the text of the message unaltered for now.

So, now it looks like this, at the top of the screen:
image

I will add sound support if someone tells me what sound to play and what volume to use (i.e. what lua snippet for sound playing I should use)

File: maps/biter_battles_v2/main.lua [hunk 1]
```diff
@@ -81,9 +83,56 @@ local function on_research_reversed(event)
...
+					end
+				end
+				table.remove(global.pings_to_remove, 1)
+			else
+				break
+			end
+		end
+	end)
+
+local function do_ping(to_player, from_player_name, message)
```
Comment (penguinencounter) [inline maps/biter_battles_v2/main.lua:106]: This function is called inconsistently:
* in `Public.print_message_to_players`: `ping_fn(playerOfForce, playerNameSendingMessage, msgToPrint)`
* in `on_console_chat`: `ping_fn_map[pinged_name](player, event.message)`

Calls from `on_console_chat` miss `from_player_name` and provide the message instead, leaving `message` nil and a blank space in the ping UI.

Also, `on_console_chat` invocations don't include the sending player username in the message.

**Review by penguinencounter — CHANGES_REQUESTED** [review]
* there is no maximum size of the window. The alignment code seems to assume a 400-unit wide dialog, so perhaps setting `width = 400` would work
long message. notification window is overflowing off the screen

File: maps/biter_battles_v2/main.lua (hunk 1 above)
Comment (clifffrey) [reply maps/biter_battles_v2/main.lua]: Fixed.

**Comment by developer-8** [comment]
- just message to some team triggers ping 
 image
- /whisper, /reply, /r - not ping
- additional message sound to standard message sound  and damage are excessive with popup

**Review by developer-8 — APPROVED** [review]

**Comment by developer-8** [comment]
might need `/ignore-ping username` command to block pings from trolls, but it can be done in another PR

**Pull request merged**